### ISSUE TITLE
Avoid trying to do locking on NFS (and BeeGFS) by reorganizing Toil's directories

### ISSUE THREAD

**Issue by gmloose** [DataBiosphere/toil#3880]
I've run into several issues with Toil lately, when using a shared working directory on an NFS file system (see issue #3867). File not found errors, stale file handles, non-empty directories that should have been empty, etc. A colleague of mine ran into similar issues on his cluster, which uses the BeeGFS file system.

When looking at the different tracebacks, I cannot conclude anything different than that there's a classical race condition. After a bit of googling, I learned that file locking does _not_ work reliably on NFS file systems. See for example:
- https://stackoverflow.com/a/38239456/16350552, and 
- https://chris.improbable.org/2010/12/16/everything-you-never-wanted-to-know-about-file-locking/

It seems that the only sensible solution is to change the way file locking is done at the moment. The open(2) man page hints at a solution (see section on `O_EXCL`):

              Portable programs that want to perform atomic file locking
              using a lockfile, and need to avoid reliance on NFS
              support for O_EXCL, can create a unique file on the same
              filesystem (e.g., incorporating hostname and PID), and use
              link(2) to make a link to the lockfile.  If link(2)
              returns 0, the lock is successful.  Otherwise, use stat(2)
              on the unique file to check if its link count has
              increased to 2, in which case the lock is also successful.

There are a couple of python packages that seem to provide a reliable file locking solution:
- [filelock](https://py-filelock.readthedocs.io/en/latest/index.html)
- [flufl.lock](https://flufllock.readthedocs.io/en/stable/)
- [portalocker](https://github.com/WoLpH/portalocker)

You might also consider a do-it-yourself solution, but that has the likely drawback of being less well-tested than existing solutions.

┆Issue is synchronized with this [Jira Story](https://ucsc-cgl.atlassian.net/browse/TOIL-1071)
┆friendlyId: TOIL-1071

**Comment by adamnovak** [comment]
One important consideration is that we rely on locks going away when the holding process dies, if it dies while holding the lock. So we'd need to use a library that can provide that.

Toil doesn't actually need or really actually want any of the directories where we do file locking to be shared across machines, though. We'd much prefer them to be on local scratch space not backed by NFS. Even when they are on NFS, we do, IIRC, all of our locking inside per-machine directories, because we specifically don't want to be trying to do distributed locking.

@gmloose Why are you using an NFS-backed `--workDir`? Can't you use a POSIX-conformant filesystem local to each node instead? Do your nodes not have any or enough local POSIX-conformant scratch space? Should Toil wander the filesystem a bit to try and find some writable local disk space instead of whatever default path it is using, which for you happens to be on NFS? Should Toil maybe keep all its per-machine locking and deferred function stuff in memory at `/var/run/user/<uid>/toil/<workflow>` instead of on disk at all?

**Comment by gmloose** [comment]
@adamnovak The main reason for doing so, IIRC, was that the documentation for `--workDir` says that if you want to be able to access `stdout` and `stderr` of your job you need to have the working directory on a shared file system. 

Here's the quote from https://toil.readthedocs.io/en/latest/running/hpcEnvironments.html?highlight=workDir#standard-output-error-from-batch-system-jobs
> If capturing standard output and error is desired, --workDir will generally need to be on a shared file system; otherwise if these are written to local temporary directories on each node (e.g. /tmp) Toil will not be able to retrieve them. 

To be honest I'm a bit confused by this, because the documentation for the command-line option `--writeLogs` (https://toil.readthedocs.io/en/latest/running/cliOptions.html?highlight=writeLogs#optionsref) says:
> ... Any non-empty standard output and error from failed batch system jobs will also be written into files at this path. ...

So, my take on this now, by re-reading the docs, combined with your question is that, _if_ you specify `--writeLogs` then, even if `--workDir` points to local scratch space (or is not specified at all), you will still get the `stdout` and `stderr` from a failed job. I think that will generally suffice for debugging purposes.

So, I did some testing yesterday evening. If I remove the `--workDir` option from the command line, then all the NFS-related errors are indeed gone. I still need to use `--temp-outdir-prefix`, because each compute node has its own `/tmp` directory, and `--writeLogs` to ensure that logs of failed jobs are retained. But, I guess there's no file locking happening is those directories, right?

I like your suggestion to keep all its per-machine locking and deferred function stuff in `/var/run`. That way, even if `workDir` is pointing to a shared disk that is not fully POSIX-compliant, you won't run into file locking race conditions.

**Comment by adamnovak** [comment]
So maybe what we want to do is:

1. Move as much stuff we need to lock as possible into /var/run (unless we're on MacOS) so it will Just Work. We might or might not be able to do this with e.g. the SQLite database for caching, and maybe someone is keeping huge deferred functions around, but I actually think it would work for nearly everything.
2. Move the HPC batch engine log dump location (default `--writeLogs`) out of `--workDir` to somewhere we actually would expect to be shared. The docs say it defaults to the current "working" directory (i.e. where the workflow is run from), which you would think would be fine, but it seems like the docs may be lying if you *need* a `--writeLogs` for this to work.
3. Maybe rename `--workDir` to `--eachNodeDir` or something to make it clear that we would really like it to be node-local.

@gmloose Are you using `--bypass-file-store` and that's why you need `--tmp-outdir-prefix` to be a shared directory? CWL workflows shouldn't need any shared storage other than the job store anymore in the default mode.

**Comment by gmloose** [comment]
Yes, I'm using `--bypass-file-store`, because some of the workflows use `InplaceUpdateRequirement`. If I don't add this option, I get the following error (which, by the way is very informative): 
```
cwltool.errors.UnsupportedRequirement: Toil cannot support InplaceUpdateRequirement when using the Toil file store. If you are running on a single machine, or a cluster with a shared filesystem, use the --bypass-file-store option to keep intermediate files on the filesystem. You can use the --tmp-outdir-prefix, --tmpdir-prefix, --outdir, and --jobStore options to control where on the filesystem files are placed, if only some parts of the filesystem are shared.
```
I tried removing `--tmp-outdir-prefix` from the command line, (with `--bypass-file-store` still there), but then my workflow fails miserably.

Ad 2) The reason I use `--writeLogs` is that I want to have control over where the logs are written. Otherwise, it might be hard to find them, unless you know exactly from which directory the workflow was run.

Ad 3) I find the name `--eachNodeDir` rather confusing. Maybe something like `--localDir`, or `--localScratchDir`, or `--bookKeepingDir`? Would you still need `--workDir` at all, if you would place all internal book-keeping files into `/var/run`?

**Comment by adamnovak** [comment]
I fixed point 1 here in #4059, and I just split off point 2, so I am going to close this. We should now keep the things that need locking in /var/run, and it should be safer to keep `--workDir` on NFS. We still might get stuck leaving behind trash that NFS won't let us delete, but we shouldn't have any correctness problems with file locking anymore.
